### PR TITLE
Separate module schema from implementation of functions in kmedoids_pywrapper.cpp

### PR DESCRIPTION
**Pull request by mailology**
I have made the following changes:

- Create header`kmedoids_pywrapper.hpp` for the `KMedoidsWrapper` class. The class methods are defined in different `.cpp` files with one for each file for speed up:
  1. `fitPython()`: `fit_python.cpp`
  2. `getMedoidsFinalPython()` : `medoids_python.cpp`
  3. `getMedoidsBuildPython()`: `build_medoids_python.cpp`
  4. `getLabelsPython()`: `labels_python.cpp`
  5. `getStepsPython()`: `steps_python.cpp`

- Define functions correspond to the KMedoidsWrapper class methods for Python bindings. For example: class method `fitPython()` is binded in `pybind11_module` by the function `void fit_python(py::class_<KMedoidsWrapper> &)`
- Define py::class object `cls` as `py::class_<KMedoidsWrapper> cls(m, "KMedoids")` so that it is used as an argument in the new functions (the style for binding other functions is slightly modified  then).

The updated code is tested on Python using MNIST-1k dataset. All algorithms give expected results.

### CODE REVIEW

File: headers/kmedoids_pywrapper.hpp [hunk 1]
```diff
@@ -0,0 +1,81 @@
+#ifndef FITPYTHON_H_
+#define FITPYTHON_H_
```
Comment (motiwari) [inline headers/kmedoids_pywrapper.hpp:2]: Change to `KMEDOIDS_PYWRAPPER_H_`

File: headers/kmedoids_pywrapper.hpp [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+
+void medoids_python(py::class_<KMedoidsWrapper> &);
+
+void build_medoids_python(py::class_<KMedoidsWrapper> &);
+
+void labels_python(py::class_<KMedoidsWrapper> &);
+
+void steps_python(py::class_<KMedoidsWrapper> &);
+
+# endif // FITPYTHON_H_
```
Comment (motiwari) [inline headers/kmedoids_pywrapper.hpp:81]: same

File: src/fit_python.cpp [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+                                py::kwargs kw
+) 
+{
+    // throw an error if the number of medoids is not specified in either 
+    // the KMedoids object or the fitPython function
+    try {
+        if (KMedoids::getNMedoids() == 0) { // Check for 0 as NULL
+        if (kw.size() == 0) {
+            throw py::value_error("Must specify number of medoids via n_medoids in KMedoids or k in fit function.");
+        }
```
Comment (motiwari) [inline src/fit_python.cpp:42]: indent

File: src/kmedoids_pywrapper.cpp [hunk 4]
```diff
@@ -38,116 +39,34 @@ int sum_thread_ids() {
...
-      .def_property("buildConfidence", &KMedoidsWrapper::getbuildConfidence, &KMedoidsWrapper::setbuildConfidence)
-      .def_property("swapConfidence", &KMedoidsWrapper::getswapConfidence, &KMedoidsWrapper::setswapConfidence)
-      .def_property("logFilename", &KMedoidsWrapper::getLogfileName, &KMedoidsWrapper::setLogFilename)
-      .def_property_readonly("medoids", &KMedoidsWrapper::getMedoidsFinalPython)
-      .def_property_readonly("build_medoids", &KMedoidsWrapper::getMedoidsBuildPython)
-      .def_property_readonly("labels", &KMedoidsWrapper::getLabelsPython)
-      .def_property_readonly("steps", &KMedoidsWrapper::getStepsPython)
-      .def("fit", &KMedoidsWrapper::fitPython);
+  py::class_<KMedoidsWrapper> cls(m, "KMedoids"); 
+  cls.def(py::init<int, std::string, int, int, int, int, std::string>(),
```
Comment (motiwari) [inline src/kmedoids_pywrapper.cpp:49]: good changes! much more readable

File: src/kmedoids_pywrapper.cpp [hunk 5]
```diff
@@ -38,116 +39,34 @@ int sum_thread_ids() {
...
+  cls.def_property("verbosity", &KMedoidsWrapper::getVerbosity, &KMedoidsWrapper::setVerbosity);
+  cls.def_property("maxIter", &KMedoidsWrapper::getMaxIter, &KMedoidsWrapper::setMaxIter);
+  cls.def_property("buildConfidence", &KMedoidsWrapper::getbuildConfidence, &KMedoidsWrapper::setbuildConfidence);
+  cls.def_property("swapConfidence", &KMedoidsWrapper::getswapConfidence, &KMedoidsWrapper::setswapConfidence);
+  cls.def_property("logFilename", &KMedoidsWrapper::getLogfileName, &KMedoidsWrapper::setLogFilename);
+  medoids_python(cls);
+  build_medoids_python(cls);
+  labels_python(cls);
+  steps_python(cls);
+  fit_python(cls);
```
Comment (motiwari) [inline src/kmedoids_pywrapper.cpp:69]: Can you educate me about the syntax of these lines? Why not include the `cls.def_property_readonly`? And how are these linked to their C++ fns? I.e. `&KMedoidsWrapper::getLabelsPython` for `labels`?

**Review by motiwari — COMMENTED** [review]
Nits and questions

**Review by motiwari — CHANGES_REQUESTED** [review]
Updating review

File: src/kmedoids_pywrapper.cpp (hunk 5 above)
Comment (mailology) [reply src/kmedoids_pywrapper.cpp:69]: This is the example I follow: https://pybind11.readthedocs.io/en/stable/faq.html?highlight=reduce%20build%20time#how-can-i-reduce-the-build-time

Take `labels` as an example:

In `labels_python.cpp`, I have defined two functions `&KMedoidsWrapper::getLabelsPython()` and `labels_python()`. Essentially, `labels_python()` applies the `cls.def_property_readonly` to `&KMedoidsWrapper::getLabelsPython()`. That's why we only need to call this function in the binding! Note that all the new functions are defined in the header `kmedoids_pywrapper.hpp`.

**Comment by mailology** [comment]
Fixed the nits. I have made some comments on the questions asked above. The updated code gives expected results on Python implementation.

**Review by motiwari — APPROVED** [review]
lgtm, thanks @mailology ! can you rebase/fix the merge conflicts?

**Pull request merged**